router/core/header_rule_engine.go

```diff
@@ -324,31 +324,34 @@ func (h *HeaderPropagation) applyResponseRule(propagation *responseHeaderPropaga
 					if slices.Contains(ignoredHeaders, name) {
 						continue
 					}
-					h.applyResponseRuleKeyValue(res, propagation, rule, name, res.Header.Get(name))
+					values := res.Header.Values(name)
+					h.applyResponseRuleKeyValue(res, propagation, rule, name, values)
 				}
 			}
 		}
 	} else if rule.Algorithm == config.ResponseHeaderRuleAlgorithmMostRestrictiveCacheControl {
 		// Explicitly apply the CacheControl algorithm on the headers
-		h.applyResponseRuleKeyValue(res, propagation, rule, "", "")
+		h.applyResponseRuleKeyValue(res, propagation, rule, "", []string{""})
 	}
 }
 
-func (h *HeaderPropagation) applyResponseRuleKeyValue(res *http.Response, propagation *responseHeaderPropagation, rule *config.ResponseHeaderRule, key, value string) {
+func (h *HeaderPropagation) applyResponseRuleKeyValue(res *http.Response, propagation *responseHeaderPropagation, rule *config.ResponseHeaderRule, key string, values []string) {
+	// Since we'll be setting the header map directly, we need to canonicalize the key
+	key = http.CanonicalHeaderKey(key)
 	switch rule.Algorithm {
 	case config.ResponseHeaderRuleAlgorithmFirstWrite:
 		propagation.m.Lock()
 		if val := propagation.header.Get(key); val == "" {
-			propagation.header.Set(key, value)
+			propagation.header[key] = values
```

Comment:
@cmtm apologize, maybe I confused you by commenting code in tests

the main problem here is that this change has different behavior than it was before
this was applying the canonical form of a header name
```
propagation.header.Set(key, value)
```
this one
```
propagation.header[key] = values
```

is not doing that, because you manually modify the header map

Reply:
@devsergiy, does this make sense? We should still be canonicalizing it.

Reply:
yep it is fine now
sorry if I missed it earlier, hard to track changes with the force pushes